File: core/src/schema.rs
use crate::connection::{connect_and_bind, ConnectionError};
use ldap3::{Scope, SearchEntry};

#[derive(uniffi::Record, Clone, Default)]
pub struct SchemaObjectClass {
    pub oid: String,
    /// All NAME values — the first is the conventional display name, any
    /// further ones are aliases.
    pub names: Vec<String>,
    pub description: Option<String>,
    pub obsolete: bool,
    pub superior_classes: Vec<String>,
    /// STRUCTURAL / ABSTRACT / AUXILIARY — defaults to STRUCTURAL per RFC
    /// 4512 when the server omits it.
    pub kind: String,
    pub must: Vec<String>,
    pub may: Vec<String>,
    pub x_origin: Option<String>,
    /// The untouched schema definition string, for anyone who wants to see
    /// exactly what the server sent regardless of how we parsed it.
    pub raw: String,
}

#[derive(uniffi::Record, Clone, Default)]
pub struct SchemaAttributeType {
    pub oid: String,
    pub names: Vec<String>,
    pub description: Option<String>,
    pub obsolete: bool,
    pub superior_type: Option<String>,
    pub equality_matching_rule: Option<String>,
    pub ordering_matching_rule: Option<String>,
    pub substring_matching_rule: Option<String>,
    /// The SYNTAX token as the server sent it — usually a numeric OID,
    /// sometimes with a `{length}` suffix.
    pub syntax_oid: Option<String>,
    pub single_valued: bool,
    pub collective: bool,
    pub no_user_modification: bool,
    pub usage: Option<String>,
    pub x_origin: Option<String>,
    pub raw: String,
}

#[derive(uniffi::Record, Clone, Default)]
pub struct LdapSchema {
    pub object_classes: Vec<SchemaObjectClass>,
    pub attribute_types: Vec<SchemaAttributeType>,
}

/// A small hand-rolled reader for RFC 4512's schema description grammar —
/// not a general XML/JSON-style parser, just enough to walk the specific
/// `( oid [ NAME ... ] [ DESC ... ] ... )` shape objectClass and
/// attributeType descriptions always use.
struct Scanner<'a> {
    chars: std::iter::Peekable<std::str::Chars<'a>>,
}

impl<'a> Scanner<'a> {
    fn new(s: &'a str) -> Self {
        Scanner { chars: s.chars().peekable() }
    }

    fn skip_ws(&mut self) {
        while matches!(self.chars.peek(), Some(c) if c.is_whitespace()) {
            self.chars.next();
        }
    }

    fn peek(&mut self) -> Option<char> {
        self.chars.peek().copied()
    }

    /// A bare token (OID, descriptor, or keyword) — anything up to the next
    /// whitespace, parenthesis, or `$`.
    fn read_word(&mut self) -> Option<String> {
        self.skip_ws();
        let mut result = String::new();
        while let Some(&c) = self.chars.peek() {
            if c.is_whitespace() || c == '(' || c == ')' || c == '$' {
                break;
            }
            result.push(c);
            self.chars.next();
        }
        if result.is_empty() { None } else { Some(result) }
    }

    /// A single-quoted string, unescaping `\XX` hex-byte escapes.
    fn read_quoted(&mut self) -> Option<String> {
        self.skip_ws();
        if self.chars.peek() != Some(&'\'') {
            return None;
        }
        self.chars.next();
        let mut result = String::new();
        while let Some(c) = self.chars.next() {
            if c == '\'' {
                return Some(result);
            }
            if c == '\\' {
                let (h1, h2) = (self.chars.next(), self.chars.next());
                if let (Some(h1), Some(h2)) = (h1, h2) {
                    if let Ok(byte) = u8::from_str_radix(&format!("{h1}{h2}"), 16) {
                        result.push(byte as char);
                        continue;
                    }
                    result.push(c);
                    result.push(h1);
                    result.push(h2);
                    continue;
                }
            }
            result.push(c);
        }
        Some(result)
    }

    /// Either one bare token, or a parenthesized `$`-separated list of them
    /// — used for SUP/MUST/MAY, which can each be single or multi-valued.
    fn read_oid_list(&mut self) -> Vec<String> {
        self.skip_ws();
        if self.peek() != Some('(') {
            return self.read_word().map(|w| vec![w]).unwrap_or_default();
        }
        self.chars.next();
        let mut items = Vec::new();
        loop {
            self.skip_ws();
            match self.peek() {
                Some(')') => {
                    self.chars.next();
                    break;
                }
                Some('$') => {
                    self.chars.next();
                }
                None => break,
                _ => match self.read_word() {
                    Some(word) => items.push(word),
                    None => break,
                },
            }
        }
        items
    }

    /// Either one quoted string, or a parenthesized list of them — used for
    /// NAME and for extension (X-...) values.
    fn read_qdescrs(&mut self) -> Vec<String> {
        self.skip_ws();
        if self.peek() != Some('(') {
            return self.read_quoted().map(|s| vec![s]).unwrap_or_default();
        }
        self.chars.next();
        let mut items = Vec::new();
        loop {
            self.skip_ws();
            match self.peek() {
                Some(')') => {
                    self.chars.next();
                    break;
                }
                None => break,
                _ => match self.read_quoted() {
                    Some(s) => items.push(s),
                    None => break,
                },
            }
        }
        items
    }
}

fn parse_object_class(raw: &str) -> Option<SchemaObjectClass> {
    let mut s = Scanner::new(raw.trim());
    s.skip_ws();
    if s.peek() != Some('(') {
        return None;
    }
    s.chars.next();

    let oid = s.read_word()?;
    let mut result = SchemaObjectClass {
        oid,
        kind: "STRUCTURAL".to_string(),
        raw: raw.to_string(),
        ..Default::default()
    };

    loop {
        s.skip_ws();
        match s.peek() {
            Some(')') => {
                s.chars.next();
                break;
            }
            None => break,
            _ => {}
        }
        let Some(keyword) = s.read_word() else { break };
        match keyword.to_uppercase().as_str() {
            "NAME" => result.names = s.read_qdescrs(),
            "DESC" => result.description = s.read_quoted(),
            "OBSOLETE" => result.obsolete = true,
            "SUP" => result.superior_classes = s.read_oid_list(),
            "STRUCTURAL" | "ABSTRACT" | "AUXILIARY" => result.kind = keyword.to_uppercase(),
            "MUST" => result.must = s.read_oid_list(),
            "MAY" => result.may = s.read_oid_list(),
            other if other.to_uppercase().starts_with("X-") => {
                let values = s.read_qdescrs();
                if other.eq_ignore_ascii_case("X-ORIGIN") {
                    result.x_origin = values.first().cloned();
                }
            }
            _ => {}
        }
    }

    Some(result)
}

fn parse_attribute_type(raw: &str) -> Option<SchemaAttributeType> {
    let mut s = Scanner::new(raw.trim());
    s.skip_ws();
    if s.peek() != Some('(') {
        return None;
    }
    s.chars.next();

    let oid = s.read_word()?;
    let mut result = SchemaAttributeType {
        oid,
        raw: raw.to_string(),
        ..Default::default()
    };

    loop {
        s.skip_ws();
        match s.peek() {
            Some(')') => {
                s.chars.next();
                break;
            }
            None => break,
            _ => {}
        }
        let Some(keyword) = s.read_word() else { break };
        match keyword.to_uppercase().as_str() {
            "NAME" => result.names = s.read_qdescrs(),
            "DESC" => result.description = s.read_quoted(),
            "OBSOLETE" => result.obsolete = true,
            "SUP" => result.superior_type = s.read_word(),
            "EQUALITY" => result.equality_matching_rule = s.read_word(),
            "ORDERING" => result.ordering_matching_rule = s.read_word(),
            "SUBSTR" => result.substring_matching_rule = s.read_word(),
            "SYNTAX" => result.syntax_oid = s.read_word(),
            "SINGLE-VALUE" => result.single_valued = true,
            "COLLECTIVE" => result.collective = true,
            "NO-USER-MODIFICATION" => result.no_user_modification = true,
            "USAGE" => result.usage = s.read_word(),
            other if other.to_uppercase().starts_with("X-") => {
                let values = s.read_qdescrs();
                if other.eq_ignore_ascii_case("X-ORIGIN") {
                    result.x_origin = values.first().cloned();
                }
            }
            _ => {}
        }
    }

    Some(result)
}

/// Fetches and parses the server's schema — object classes and attribute
/// types. The schema isn't at a fixed, well-known DN; the Root DSE's
/// `subschemaSubentry` operational attribute tells us where to look for it,
/// per RFC 4512.
#[uniffi::export(async_runtime = "tokio")]
pub async fn fetch_schema(
    host: String,
    port: u16,
    use_ssl: bool,
    bind_dn: String,
    password: String,
) -> Result<LdapSchema, ConnectionError> {
    let mut ldap = connect_and_bind(&host, port, use_ssl, &bind_dn, &password).await?;

    let (root_dse_results, _) = ldap
        .search("", Scope::Base, "(objectClass=*)", vec!["subschemaSubentry"])
        .await
        .map_err(|e| ConnectionError::SearchFailed { reason: e.to_string() })?
        .success()
        .map_err(|e| ConnectionError::SearchFailed { reason: e.to_string() })?;

    let subschema_dn = root_dse_results
        .into_iter()
        .next()
        .map(SearchEntry::construct)
        .and_then(|entry| entry.attrs.get("subschemaSubentry").and_then(|v| v.first().cloned()))
        .unwrap_or_else(|| "cn=subschema".to_string());

    let (schema_results, _) = ldap
        .search(&subschema_dn, Scope::Base, "(objectClass=*)", vec!["objectClasses", "attributeTypes"])
        .await
        .map_err(|e| ConnectionError::SearchFailed { reason: e.to_string() })?
        .success()
        .map_err(|e| ConnectionError::SearchFailed { reason: e.to_string() })?;

    let _ = ldap.unbind().await;

    let entry = schema_results
        .into_iter()
        .next()
        .map(SearchEntry::construct)
        .ok_or_else(|| ConnectionError::SearchFailed {
            reason: format!("Schema entry \"{subschema_dn}\" was not found"),
        })?;

    let object_classes = entry
        .attrs
        .get("objectClasses")
        .map(|values| values.iter().filter_map(|v| parse_object_class(v)).collect())
        .unwrap_or_default();

    let attribute_types = entry
        .attrs
        .get("attributeTypes")
        .map(|values| values.iter().filter_map(|v| parse_attribute_type(v)).collect())
        .unwrap_or_default();

    Ok(LdapSchema { object_classes, attribute_types })
}

File: core/src/schema.c
/*
 * schema.c — ls_fetch_schema: fetch and parse the server's objectClasses
 * and attributeTypes. A direct port of the old Rust schema.rs, including
 * its small hand-rolled RFC 4512 scanner.
 *
 * The schema isn't at a fixed DN: the Root DSE's subschemaSubentry
 * operational attribute points at it (RFC 4512), with cn=subschema as the
 * conventional fallback.
 */
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h> /* strcasecmp */

#include "internal_ldap.h"

/* ── scanner ─────────────────────────────────────────────────────── */

typedef struct {
  const char *p;
  const char *end;
} Scanner;

static void sc_init(Scanner *s, const char *str) {
  s->p = str;
  s->end = str + strlen(str);
}

static void sc_skip_ws(Scanner *s) {
  while (s->p < s->end && isspace((unsigned char)*s->p)) s->p++;
}

static char sc_peek(Scanner *s) { return s->p < s->end ? *s->p : '\0'; }

/* Bare token: up to the next whitespace, parenthesis, or '$'. NULL if empty. */
static char *sc_read_word(Scanner *s) {
  sc_skip_ws(s);
  const char *start = s->p;
  while (s->p < s->end) {
    char c = *s->p;
    if (isspace((unsigned char)c) || c == '(' || c == ')' || c == '$') break;
    s->p++;
  }
  if (s->p == start) return NULL;
  return ls_strdup_n(start, (size_t)(s->p - start));
}

/* Single-quoted string, unescaping "\XX" hex-byte escapes. */
static char *sc_read_quoted(Scanner *s) {
  sc_skip_ws(s);
  if (sc_peek(s) != '\'') return NULL;
  s->p++;

  size_t cap = 16;
  size_t n = 0;
  char *out = ls_xmalloc(cap);
  while (s->p < s->end) {
    char c = *s->p++;
    if (c == '\'') break;
    if (c == '\\' && s->p + 1 < s->end) {
      char h1 = s->p[0];
      char h2 = s->p[1];
      char hex[3] = {h1, h2, 0};
      char *endp = NULL;
      long byte = strtol(hex, &endp, 16);
      if (endp == hex + 2) {
        s->p += 2;
        c = (char)byte;
      }
    }
    if (n + 1 >= cap) {
      cap *= 2;
      out = ls_xrealloc(out, cap);
    }
    out[n++] = c;
  }
  out[n] = '\0';
  return out;
}

/* One bare token, or a parenthesized '$'-separated list (SUP/MUST/MAY). */
static void sc_read_oid_list(Scanner *s, LSStrList *list) {
  ls_strlist_init(list);
  sc_skip_ws(s);
  if (sc_peek(s) != '(') {
    char *w = sc_read_word(s);
    if (w) ls_strlist_push(list, w);
    return;
  }
  s->p++;
  for (;;) {
    sc_skip_ws(s);
    char c = sc_peek(s);
    if (c == ')') {
      s->p++;
      break;
    }
    if (c == '$') {
      s->p++;
      continue;
    }
    if (c == '\0') break;
    char *w = sc_read_word(s);
    if (!w) break;
    ls_strlist_push(list, w);
  }
}

/* One quoted string, or a parenthesized list of them (NAME, X-... values). */
static void sc_read_qdescrs(Scanner *s, LSStrList *list) {
  ls_strlist_init(list);
  sc_skip_ws(s);
  if (sc_peek(s) != '(') {
    char *q = sc_read_quoted(s);
    if (q) ls_strlist_push(list, q);
    return;
  }
  s->p++;
  for (;;) {
    sc_skip_ws(s);
    char c = sc_peek(s);
    if (c == ')') {
      s->p++;
      break;
    }
    if (c == '\0') break;
    char *q = sc_read_quoted(s);
    if (!q) break;
    ls_strlist_push(list, q);
  }
}

static void upper_ascii(char *s) {
  for (; *s; ++s) *s = (char)toupper((unsigned char)*s);
}

static char *first_or_null(LSStrList *l) {
  if (l->count == 0) {
    ls_strlist_free_items(l);
    return NULL;
  }
  char *first = ls_xstrdup(l->items[0]);
  ls_strlist_free_items(l);
  return first;
}

/* ── parsers ─────────────────────────────────────────────────────── */

static bool parse_object_class(const char *raw, LSObjectClass *out) {
  Scanner s;
  sc_init(&s, raw);
  sc_skip_ws(&s);
  if (sc_peek(&s) != '(') return false;
  s.p++;

  char *oid = sc_read_word(&s);
  if (!oid) return false;

  memset(out, 0, sizeof *out);
  out->oid = oid;
  out->kind = ls_xstrdup("STRUCTURAL");
  out->raw = ls_xstrdup(raw);

  for (;;) {
    sc_skip_ws(&s);
    char c = sc_peek(&s);
    if (c == ')') {
      s.p++;
      break;
    }
    if (c == '\0') break;

    char *kw = sc_read_word(&s);
    if (!kw) break;
    char up[64];
    snprintf(up, sizeof up, "%s", kw);
    upper_ascii(up);

    if (strcmp(up, "NAME") == 0) {
      LSStrList l;
      sc_read_qdescrs(&s, &l);
      out->names = l.items;
      out->name_count = l.count;
    } else if (strcmp(up, "DESC") == 0) {
      free(out->description);
      out->description = sc_read_quoted(&s);
    } else if (strcmp(up, "OBSOLETE") == 0) {
      out->obsolete = true;
    } else if (strcmp(up, "SUP") == 0) {
      LSStrList l;
      sc_read_oid_list(&s, &l);
      out->superior_classes = l.items;
      out->superior_class_count = l.count;
    } else if (strcmp(up, "STRUCTURAL") == 0 || strcmp(up, "ABSTRACT") == 0 ||
               strcmp(up, "AUXILIARY") == 0) {
      free(out->kind);
      out->kind = ls_xstrdup(up);
    } else if (strcmp(up, "MUST") == 0) {
      LSStrList l;
      sc_read_oid_list(&s, &l);
      out->must = l.items;
      out->must_count = l.count;
    } else if (strcmp(up, "MAY") == 0) {
      LSStrList l;
      sc_read_oid_list(&s, &l);
      out->may = l.items;
      out->may_count = l.count;
    } else if (strncmp(up, "X-", 2) == 0) {
      LSStrList l;
      sc_read_qdescrs(&s, &l);
      if (strcasecmp(kw, "X-ORIGIN") == 0 && !out->x_origin)
        out->x_origin = first_or_null(&l);
      else
        ls_strlist_free_items(&l);
    }
    free(kw);
  }
  return true;
}

static bool parse_attribute_type(const char *raw, LSAttributeType *out) {
  Scanner s;
  sc_init(&s, raw);
  sc_skip_ws(&s);
  if (sc_peek(&s) != '(') return false;
  s.p++;

  char *oid = sc_read_word(&s);
  if (!oid) return false;

  memset(out, 0, sizeof *out);
  out->oid = oid;
  out->raw = ls_xstrdup(raw);

  for (;;) {
    sc_skip_ws(&s);
    char c = sc_peek(&s);
    if (c == ')') {
      s.p++;
      break;
    }
    if (c == '\0') break;

    char *kw = sc_read_word(&s);
    if (!kw) break;
    char up[64];
    snprintf(up, sizeof up, "%s", kw);
    upper_ascii(up);

    if (strcmp(up, "NAME") == 0) {
      LSStrList l;
      sc_read_qdescrs(&s, &l);
      out->names = l.items;
      out->name_count = l.count;
    } else if (strcmp(up, "DESC") == 0) {
      free(out->description);
      out->description = sc_read_quoted(&s);
    } else if (strcmp(up, "OBSOLETE") == 0) {
      out->obsolete = true;
    } else if (strcmp(up, "SUP") == 0) {
      free(out->superior_type);
      out->superior_type = sc_read_word(&s);
    } else if (strcmp(up, "EQUALITY") == 0) {
      free(out->equality_matching_rule);
      out->equality_matching_rule = sc_read_word(&s);
    } else if (strcmp(up, "ORDERING") == 0) {
      free(out->ordering_matching_rule);
      out->ordering_matching_rule = sc_read_word(&s);
    } else if (strcmp(up, "SUBSTR") == 0) {
      free(out->substring_matching_rule);
      out->substring_matching_rule = sc_read_word(&s);
    } else if (strcmp(up, "SYNTAX") == 0) {
      free(out->syntax_oid);
      out->syntax_oid = sc_read_word(&s);
    } else if (strcmp(up, "SINGLE-VALUE") == 0) {
      out->single_valued = true;
    } else if (strcmp(up, "COLLECTIVE") == 0) {
      out->collective = true;
    } else if (strcmp(up, "NO-USER-MODIFICATION") == 0) {
      out->no_user_modification = true;
    } else if (strcmp(up, "USAGE") == 0) {
      free(out->usage);
      out->usage = sc_read_word(&s);
    } else if (strncmp(up, "X-", 2) == 0) {
      LSStrList l;
      sc_read_qdescrs(&s, &l);
      if (strcasecmp(kw, "X-ORIGIN") == 0 && !out->x_origin)
        out->x_origin = first_or_null(&l);
      else
        ls_strlist_free_items(&l);
    }
    free(kw);
  }
  return true;
}

/* ── fetch ───────────────────────────────────────────────────────── */

/* First value of `attr` on the first entry of `res`, as an owned string. */
static char *first_value(LDAP *ld, LDAPMessage *res, const char *attr) {
  LDAPMessage *e = ldap_first_entry(ld, res);
  if (!e) return NULL;
  struct berval **vals = ldap_get_values_len(ld, e, attr);
  if (!vals || !vals[0]) {
    if (vals) ldap_value_free_len(vals);
    return NULL;
  }
  char *out = ls_strdup_n(vals[0]->bv_val, vals[0]->bv_len);
  ldap_value_free_len(vals);
  return out;
}

int ls_fetch_schema(const char *host, uint16_t port, bool use_ssl,
                    const char *bind_dn, const char *password, LSSchema **out,
                    LSError *err) {
  *out = NULL;

  LDAP *ld = NULL;
  int rc =
      ls_connect_and_bind(host, port, use_ssl, bind_dn, password, &ld, err);
  if (rc != LS_OK) return rc;

  /* Root DSE → subschemaSubentry. */
  char *rootattrs[] = {(char *)"subschemaSubentry", NULL};
  LDAPMessage *dse = NULL;
  int lrc =
      ldap_search_ext_s(ld, "", LDAP_SCOPE_BASE, "(objectClass=*)", rootattrs,
                        0, NULL, NULL, NULL, LDAP_NO_LIMIT, &dse);
  if (lrc != LDAP_SUCCESS) {
    if (dse) ldap_msgfree(dse);
    ldap_unbind_ext_s(ld, NULL, NULL);
    return ls_ldap_fail(err, LS_SEARCH_FAILED, lrc);
  }
  char *subschema = first_value(ld, dse, "subschemaSubentry");
  ldap_msgfree(dse);
  if (!subschema) subschema = ls_xstrdup("cn=subschema");

  /* The schema entry itself. */
  char *schemaattrs[] = {(char *)"objectClasses", (char *)"attributeTypes",
                         NULL};
  LDAPMessage *sres = NULL;
  lrc =
      ldap_search_ext_s(ld, subschema, LDAP_SCOPE_BASE, "(objectClass=*)",
                        schemaattrs, 0, NULL, NULL, NULL, LDAP_NO_LIMIT, &sres);
  if (lrc != LDAP_SUCCESS) {
    if (sres) ldap_msgfree(sres);
    ldap_unbind_ext_s(ld, NULL, NULL);
    char *reason = ls_aprintf("Schema entry \"%s\" was not found", subschema);
    free(subschema);
    int r = ls_fail(err, LS_SEARCH_FAILED, NULL, 0, reason);
    free(reason);
    return r;
  }

  LDAPMessage *entry = ldap_first_entry(ld, sres);
  if (!entry) {
    char *reason = ls_aprintf("Schema entry \"%s\" was not found", subschema);
    free(subschema);
    ldap_msgfree(sres);
    ldap_unbind_ext_s(ld, NULL, NULL);
    int r = ls_fail(err, LS_SEARCH_FAILED, NULL, 0, reason);
    free(reason);
    return r;
  }
  free(subschema);

  LSSchema *schema = ls_xcalloc(1, sizeof(LSSchema));

  struct berval **ocs = ldap_get_values_len(ld, entry, "objectClasses");
  if (ocs) {
    size_t n = 0;
    while (ocs[n]) n++;
    schema->object_classes = ls_xmalloc((n ? n : 1) * sizeof(LSObjectClass));
    for (size_t i = 0; i < n; ++i) {
      char *raw = ls_strdup_n(ocs[i]->bv_val, ocs[i]->bv_len);
      if (parse_object_class(
              raw, &schema->object_classes[schema->object_class_count]))
        schema->object_class_count++;
      free(raw);
    }
    ldap_value_free_len(ocs);
  }

  struct berval **ats = ldap_get_values_len(ld, entry, "attributeTypes");
  if (ats) {
    size_t n = 0;
    while (ats[n]) n++;
    schema->attribute_types = ls_xmalloc((n ? n : 1) * sizeof(LSAttributeType));
    for (size_t i = 0; i < n; ++i) {
      char *raw = ls_strdup_n(ats[i]->bv_val, ats[i]->bv_len);
      if (parse_attribute_type(
              raw, &schema->attribute_types[schema->attribute_type_count]))
        schema->attribute_type_count++;
      free(raw);
    }
    ldap_value_free_len(ats);
  }

  ldap_msgfree(sres);
  ldap_unbind_ext_s(ld, NULL, NULL);

  *out = schema;
  return LS_OK;
}
